queue_underwriting.py

```diff
@@ -1,36 +1,71 @@
-#!/usr/bin/python3
+#!/usr/bin/env python3
 
+import yaml
 import datetime
+import calendar
 import socket
-import yaml
-
-socket_path = "/tmp/liquidsoap.sock"
-sponsors_path = '/tmp/wuvt/underwriting.yml'
+import random
 
-sponsors = yaml.safe_load(open(sponsors_path))
-now = datetime.datetime.now()
+underwritingSchedule = '/tmp/sample.yml'
+socketPath = "/tmp/liquidsoap.sock"
 
-if now.minute > 30:
-    # get underwriting for the next hour
-    now += datetime.timedelta(hours=1)
-
-if now.weekday() in sponsors and type(sponsors[now.weekday()]) == dict and \
-        now.hour in sponsors[now.weekday()]:
-    sponsor = sponsors[now.weekday()][now.hour]
-    cmd = "underwriting.push {}\n".format(sponsor).encode('utf-8')
+underwriters = yaml.safe_load(open(underwritingSchedule))
+thisHour = []
 
+def push_command(url):
+    cmd = "underwriting.push {}\n".format(url).encode('utf-8')
     sock = socket.socket(socket.AF_UNIX)
-    sock.connect(socket_path)
-
+    try:
+        sock.connect(socketPath)
+    except FileNotFoundError:
+        print("Cannot push to socket: does not exist")
+        return False
     # queue the first one immediately
     sock.send(cmd)
-
-    if now.minute > 30:
+    if datetime.datetime.now().minute > 30:
         # we must be queuing for the next hour, so add the second
         sock.send(cmd)
-
     sock.send(b'exit\n')
+    return True
+
+# check which hour we're queuing for and return datetime object
+def what_hour():
+    queue_hour = datetime.datetime.now()
+    if queue_hour.minute > 30:
+        #queuing for next hour
+        queue_hour = queue_hour + datetime.timedelta(hours=1)
+    return queue_hour
+
+queue_hour = what_hour()
+
+for underwriting in underwriters:
+    today=queue_hour.date()
+    weekday=calendar.day_name[queue_hour.weekday()].lower()
+    # we want to sanely handle do-not-kill situations
+    if ( ( not underwriting['startdate'] or underwriting['startdate'] <= today ) and\
+            ( not underwriting['enddate'] or underwriting['enddate'] >= today ) ) and\
+            weekday in underwriting['schedule'] and queue_hour.hour in underwriting['schedule'][weekday]:
+        if len(underwriting['url']) > 1:
+            # Then we probably have multiple recordings of the grant statement
+            playFile = underwriting['url'][random.randint(0,len(underwriting['url'])-1)]
+        else:
+            # we probably don't need this, since a random int between 0 and 0 is 0
+            playFile = underwriting['url'][0]
+        # probably the safer way to handle things in the future, but would require more parsing
+        #thisHour[underwriting['name']] = playFile
+        thisHour.append(playFile)
 
-    print("Underwriting queued")
+# because automation currently only plays one piece of underwriting at a time,
+# we need arbitrate which piece we push. We'll pick a random one!
+if len(thisHour) is not 0:
+    if push_command(thisHour[random.randint(0,len(thisHour)-1)]):
```

Comment:
It would be better to use `random.choice` here.

Reply:
looking at the docs, that seems more idiomatic; will do.